### PR TITLE
Add apple_metal_library rule

### LINKED ISSUE #2025: Support custom compiler options for compiling `.metal` files
To include a header inside a shader that is in a different folder one has to currently do so relatively, for example

```#include "../.../header.h"```

instead of the preferred

```#include "header_folder_path/header.h"```

This of course couples the source to the current directory structure.

@thii writes

"It's not possible with how rules_apple compiles metal files today, since they're built as part of resources so you can't pass custom metal compiler flags (i.e. for your example it'd be possible by adding -I . to the metal command). To support this, we probably need a new rule"

### PR DESCRIPTION
**Pull request by aaronsky**
Resolves #2025.

This rule should be used in situations where the default resource processing behavior is insufficient, such as when custom copts are required. It is an adaptation of the `metal_library` rule found in [rules_apple_line](https://github.com/line/rules_apple_line/blob/main/apple/metal_library.bzl), with changes made to more closely resemble `apple_intent_library`.

Here's an example:

```starlark
apple_metal_library(
    name = "metal_lib",
    copts = ["-w"],
    hdrs = ["shared.h"],
    srcs = ["shader1.metal", "shader2.metal"],
)

objc_library(
    name = "my_lib",
    ...
    data = [":metal_lib"],
)
```

This PR should be additive to how Metal is supported as data today, and should not introduce any breaking changes.

### CODE REVIEW

File: apple/internal/resource_rules/apple_metal_library.bzl [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+load(
+    "@build_bazel_rules_apple//apple/internal:platform_support.bzl",
+    "platform_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:resources.bzl",
+    "resources",
+)
+
+def _metal_apple_target_triple(platform_prerequisites):
```
Comment (luispadron) [inline apple/internal/resource_rules/apple_metal_library.bzl:47]: nit: can you add a doc comment

File: apple/internal/resource_rules/apple_metal_library.bzl [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    environment = "" if platform.is_device else "-simulator"
+
+    return "air64-apple-{platform}{version}{environment}".format(
+        environment = environment,
+        platform = platform_string,
+        version = target_os_version,
+    )
+
+def _apple_metal_library_impl(ctx):
```
Comment (luispadron) [inline apple/internal/resource_rules/apple_metal_library.bzl:63]: nit: can you add a doc comment

File: apple/internal/resource_rules/apple_metal_library.bzl [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    ]
+
+apple_metal_library = rule(
+    attrs = dicts.add(
+        apple_support.action_required_attrs(),
+        apple_toolchain_utils.shared_attrs(),
+        {
+            "copts": attr.string_list(
+                doc = """\
+A list of compiler options.
```
Comment (luispadron) [inline apple/internal/resource_rules/apple_metal_library.bzl:139]: Could we add some more details to these, maybe:

```suggestion
A list of compiler options passed to the `metal` compiler for each source.
```

File: apple/internal/resource_rules/apple_metal_library.bzl [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+""",
+            ),
+            "hdrs": attr.label_list(
+                allow_files = [".h"],
+                doc = """\
+A list of headers that you need import to metal source.
```
Comment (luispadron) [inline apple/internal/resource_rules/apple_metal_library.bzl:145]:
```suggestion
A list of headers to make importable when compiling the metal library
```

**Comment by aaronsky** [comment]
these test failures don't seem related to my changes

**Review by luispadron — APPROVED** [review]
Thanks!

File: apple/internal/resource_rules/apple_metal_library.bzl [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    target_os_version = platform_prerequisites.minimum_os
+
+    platform = platform_prerequisites.apple_fragment.single_arch_platform
+    platform_string = str(platform.platform_type)
+    if platform_string == "macos":
+        platform_string = "macosx"
+
+    environment = "" if platform.is_device else "-simulator"
+
+    return "air64-apple-{platform}{version}{environment}".format(
```
Comment (brentleyjones) [inline apple/internal/resource_rules/apple_metal_library.bzl:65]: air64?

Comment (aaronsky) [reply apple/internal/resource_rules/apple_metal_library.bzl:65]: yup! metal uses an underdocumented specialized target triple format for individual compiled shaders (`.air`) that look like LLVM triples but are decidedly not. this approach is used to harmonize better with platform transitions in rules_apple, assuming i'm understanding things correctly.

**Comment by aaronsky** [comment]
@brentleyjones is there anything else I should do before this can be merged?

**Review by brentleyjones — APPROVED** [review]

**Pull request merged**